assoc.c

```diff
@@ -292,3 +292,72 @@ void stop_assoc_maintenance_thread() {
     pthread_join(maintenance_tid, NULL);
 }
 
+struct assoc_iterator {
+    unsigned int bucket;
+    bool bucket_locked;
+    item *it;
+    item *next;
+};
+
+void *assoc_get_iterator(void) {
+    struct assoc_iterator *iter = calloc(1, sizeof(struct assoc_iterator));
+    if (iter == NULL) {
+        return NULL;
+    }
+    // this will hang the caller while a hash table expansion is running.
+    mutex_lock(&maintenance_lock);
+    return iter;
+}
+
+bool assoc_iterate(void *iterp, item **it) {
+    struct assoc_iterator *iter = (struct assoc_iterator *) iterp;
+    *it = NULL;
+    // - if locked bucket and next, update next and return
+    if (iter->bucket_locked) {
+        if (iter->next != NULL) {
+            iter->it = iter->next;
+            iter->next = iter->it->h_next;
+            *it = iter->it;
+        } else {
+            // unlock previous bucket, if any
+            item_unlock(iter->bucket);
+            // iterate the bucket post since it starts at 0.
+            iter->bucket++;
+            iter->bucket_locked = false;
+            *it = NULL;
+        }
+        return true;
+    }
+
+    // - loop until we hit the end or find something.
+    if (iter->bucket != hashsize(hashpower)) {
+        // - lock next bucket
+        item_lock(iter->bucket);
```

Comment:
We used to do `item_trylock` instead of `item_lock`. Did we change that to increase accuracy?

Reply:
LRU crawler does `item_trylock` instead of `item_lock` because the lock order is inverted: item needs to be gotten before LRU or else it can deadlock. So if you fail to get item you can't wait for it.

Since the LRU isn't involved in a hash table walk we can do a full wait. it's very similar to doing a straight GET request.

Reply:
Cool. I believe the difference (between GET) is we keep the lock until we finish walking the linked list.

Reply:
Correct. Have to hold the lock through the whole bucket, but depth will rarely be higher than 1, not more than 2-3 except for some extreme cases.

Reply:
I mean, to be specific GET does hold the lock until it finds the right item, which could be at top or bottom of the bucket :)